app/src/main/scala/org/alephium/explorer/persistence/queries/TokenQueries.scala

```diff
@@ -137,4 +137,36 @@ object TokenQueries extends StrictLogging {
       .paginate(pagination)
       .asAS[TxByTokenQR]
   }
+
+  def listAddressTokensWithBalanceAction(address: Address, pagination: Pagination)(
+      implicit ec: ExecutionContext): DBActionSR[(TokenId, U256, U256)] =
+    listAddressTokensWithBalanceUntilLockTime(address, TimeStamp.now(), pagination).map(_.map {
+      case (token, total, locked) =>
+        (token, total.getOrElse(U256.Zero), locked.getOrElse(U256.Zero))
+    })
+
+  def listAddressTokensWithBalanceUntilLockTime(
+      address: Address,
+      lockTime: TimeStamp,
+      pagination: Pagination): DBActionSR[(TokenId, Option[U256], Option[U256])] =
+    sql"""
+      SELECT
+        token_outputs.token,
+        sum(token_outputs.amount),
+        sum(CASE
+                WHEN token_outputs.lock_time is NULL or token_outputs.lock_time < ${lockTime.millis} THEN 0
+                ELSE token_outputs.amount
+            END)
+      FROM token_outputs
+               LEFT JOIN inputs
+                         ON token_outputs.key = inputs.output_ref_key
+                             AND inputs.main_chain = true
+      WHERE token_outputs.spent_finalized IS NULL
+        AND token_outputs.address = $address
+        AND token_outputs.main_chain = true
+        AND inputs.block_hash IS NULL
+      GROUP BY token_outputs.token
```

Comment:
>  it returns 0 query results for majority of addresses 

It's 0 because the `spent_finalized` field is set with `Some` here https://github.com/alephium/explorer-backend/blob/b41bed59099d1259de79a893bc58d7e08cef0ea3/app/src/test/scala/org/alephium/explorer/GenDBModel.scala#L275 I did some quick test and it's indeed rarely `None`, I should check if I could tweak it to have more even data.

I will also try to add a test to validate the balance.